Enable comparing main facets module agains sandbox facets implementation

epotyom opened the pull request:
Auxiliary changes:
- `SearchTask#getFacetResultsMsec` now includes time to run `#search`, because in the sandbox module we can't measure search and facet compute times separately. But it might be the right thing to do anyway, as otherwise we don't account for time spent to build doc ID sets in FacetsCollector?
- Enable attaching a context to Task, to be able to use different implementations for the same task and Lucene code in baseline/candidate
- Fix facets result overlap check: in python SearchTask's equals and hash methods compare facets requests, not results
- Added `facetsWikimediumAll` config that contains all taxonomy facets tasks from wikimediumall

Command to run:

```    
python src/python/localrunFacets.py -source facetsWikimediumAll
```

Results on my laptop

```
Report after iter 19:

                            TaskQPS classic_facets      StdDevQPS sandbox_facets      StdDev                Pct diff p-value
     BrowseRandomLabelTaxoFacets        5.99      (6.0%)        2.64      (1.6%)  -55.9% ( -59% -  -51%) 0.000
                        PKLookup      297.27      (5.8%)      251.45      (4.9%)  -15.4% ( -24% -   -5%) 0.000
         AndHighMedDayTaxoFacets      154.40      (3.2%)      153.00     (14.8%)   -0.9% ( -18% -   17%) 0.788
        AndHighHighDayTaxoFacets       13.13      (3.4%)       13.04     (11.7%)   -0.7% ( -15% -   14%) 0.792
          OrHighMedDayTaxoFacets       16.27      (5.5%)       19.52     (20.0%)   20.0% (  -5% -   48%) 0.000
       BrowseDayOfYearTaxoFacets        6.52      (7.7%)        9.30     (12.9%)   42.8% (  20% -   68%) 0.000
            BrowseDateTaxoFacets        6.48      (7.6%)        9.68     (13.9%)   49.3% (  25% -   76%) 0.000
           BrowseMonthTaxoFacets        6.21      (6.9%)        9.38     (15.1%)   51.0% (  27% -   78%) 0.000
            MedTermDayTaxoFacets       38.03      (3.9%)       71.68     (25.3%)   88.5% (  57% -  122%) 0.000
```

I'll look into why there is regression for BrowseRandomLabelTaxoFacets and PKLookup

---

Updates: the PR https://github.com/apache/lucene/pull/14237 that this change depends on is merged, updated test results:

```
python3.7 src/python/runFacets.py -source facetsWikimediumAll

                            TaskQPS post_collection_facets      StdDevQPS during_collection_facets      StdDev                Pct diff p-value
     BrowseRandomLabelTaxoFacets        2.45      (6.6%)        0.74      (0.2%)  -69.6% ( -71% -  -67%) 0.000
            BrowseDateTaxoFacets        3.10      (0.7%)        3.32      (2.2%)    7.2% (   4% -   10%) 0.000
       BrowseDayOfYearTaxoFacets        3.61      (1.2%)        3.98      (3.4%)   10.0% (   5% -   14%) 0.000
           BrowseMonthTaxoFacets        2.55      (3.3%)        2.82      (3.8%)   10.9% (   3% -   18%) 0.000
         AndHighMedDayTaxoFacets       77.67      (5.5%)       99.69      (2.5%)   28.3% (  19% -   38%) 0.000
          OrHighMedDayTaxoFacets        5.89      (2.1%)        7.96      (2.4%)   35.2% (  30% -   40%) 0.000
        AndHighHighDayTaxoFacets       22.99      (3.4%)       34.60      (4.1%)   50.5% (  41% -   60%) 0.000
            MedTermDayTaxoFacets       23.29      (3.1%)       49.10      (5.4%)  110.8% (  99% -  123%) 0.000
```

epotyom commented:
There is BrowseRandomLabelTaxoFacets regression because `RandomLabel.taxonomy` field is the one that uses most of the sidecar taxonomy index, and it seems to be the only field for which dense counting (in array) pays off; total array (and taxon index) size is 1559911, and when counting is done, only 3422 (0.2%) elements are still zeros.

We can think about implementing dense counting for the sandbox facet module, but it looks like a rare use case to me - not only we need a field with large number of unique values, but it also needs to be a MatchAllDocsQuery.

In any case, seems like the result is expected. Looking into PKLookup regression now.

stefanvodita commented on src/main/perf/TestContext.java:
```diff
@@ -0,0 +1,40 @@
+    public enum FacetMode {
+        UNDEFINED,
+        CLASSIC,
+        SANDBOX,  // TODO: better names?
```
Yes!

stefanvodita commented on src/python/localrunFacets.py:
```diff
@@ -0,0 +1,60 @@
+import competition
+import os
+
+# Script to compare performance of sandbox and main facets modules
```
Should we mention this file somewhere visible, maybe in the README?

epotyom replied:
Good idea, will do in the next revision.

stefanvodita reviewed (commented):
> SearchTask#getFacetResultsMsec now includes time to run #search

Will that change results we were already reporting? Just good to know.

mikemccand commented on src/main/perf/TestContext.java:
```diff
@@ -0,0 +1,40 @@
+package perf;
```
Needs copyright header?

epotyom replied:
Done!

mikemccand commented:
> * otherwise we don't account for time spent to build doc ID sets in FacetsCollector?

+1 -- good catch!

mikemccand commented on src/main/perf/SearchTask.java:
```diff
@@ -213,75 +220,190 @@ public void go(IndexState state, TaskParser taskParser) throws IOException {
+          // TODO: support sort, filter too!!
+          // TODO: support other facet methods
+          List<TaskParser.TaskBuilder.FacetTask> classicFacetRequests = new ArrayList<>();
+          List<TaskParser.TaskBuilder.FacetTask> sandboxFacetRequests = new ArrayList<>();
```
I say @epotyom and @Shradha26 should simply pick a name, now, and run with it, here :)

epotyom replied:
The way I'm thinking about it is that new classes can be moved to the existing `facets` module once we believe they are mature enough. I don't think these classes  belong in an independent module, they depend on a lot of things from the `facets` module, e.g. indexing time functionality or drill sideways. I promise to clean up and get rid of all mentions of "sandbox" once we do it, or, if decided otherwise, once we create a separate module for it. What do you think?

stefanvodita replied:
It's a good point that the new classes may become part of the facets module.
In this PR though, we're using "sanbox" as a contrast to "classic". Can we name it something generic, such as "matchTime", or are we not that detached from the particular implementation?

epotyom replied:
You have a point. How about `postCollectionFacets` for classic VS `directCollectionFacets` or `directFacets` for sandbox? Or `docIdSetFacets` VS `directFacets`?

stefanvodita replied:
I like `postCollectionFacets`. Can we pair it with `duringCollectionFacets`, `withCollectionFacets`, or `collectionFacets`?

epotyom commented:
Thank you for reviewing @stefanvodita !

> > SearchTask#getFacetResultsMsec now includes time to run #search
> 
> Will that change results we were already reporting? Just good to know.

Yeah, but TBH I can't find where we report `getFacetResultsMsec`. We print it to log, and then benchUtil reads it. Not sure what happens to it next. Maybe it is used in a different repo?

I believe that QPS metrics reported by localrun.py are not going to change.

epotyom commented on src/main/perf/SearchTask.java:
```diff
@@ -19,30 +19,40 @@
-import org.apache.lucene.facet.Facets;
-import org.apache.lucene.facet.FacetsCollector;
-import org.apache.lucene.facet.FacetsCollectorManager;
+import org.apache.lucene.facet.*;
```
Oh I think my Intellij folded it, I'll change

epotyom commented on src/python/benchUtil.py:
```diff
@@ -1193,7 +1195,9 @@ def runSimpleSearchBench(self, iter, id, c,
                 '-XX:+UnlockDiagnosticVMOptions',
-                '-XX:+DebugNonSafepoints']
+                '-XX:+DebugNonSafepoints',
+                # '-agentlib:jdwp=transport=dt_socket,server=y,suspend=y,address=localhost:7891'
```
It was intentional, it makes it easy to enable remote debug - just uncomment this line. Maybe I should add a comment. Do you think it's better to remove?

stefanvodita replied:
Personally, I don't mind leaving it in if you found it helpful for quick debugging.

stefanvodita commented on README.md:
```diff
@@ -179,3 +179,15 @@ To test vector search with [Cohere/wikipedia-22-12-en-embeddings](https://huggin
+
+There are currently two facets implementations - one that first collects document IDs and then computes facets in a separate phase, and a new implementation that computes facets during collection.
+
+To compare performance for the two implementations run
```
Should we add details about what exactly gets compared? For example, you had a comment about only taxonomy facets being compared I think.

epotyom replied:
I agree, added.

stefanvodita commented on src/main/perf/SearchTask.java:
```diff
@@ -213,75 +231,190 @@ public void go(IndexState state, TaskParser taskParser) throws IOException {
+            //       for MatchAllDocsQuery to make collection for all docs in the index faster?
+            Map<String, CountFacetRecorder> indexFieldToRecorder = new HashMap<>();
+            List<CollectorManager<? extends Collector, ?>> collectorManagers = new ArrayList<>();
+            // First collector manager in the list is to collect hits, but not for if MatchAllDocsQuery
```
Can we state this more clearly?

epotyom replied:
Reworded!

stefanvodita commented on src/main/perf/SearchTask.java:
```diff
@@ -213,75 +231,190 @@ public void go(IndexState state, TaskParser taskParser) throws IOException {
+                if (request.dimension().startsWith("range:")) {
+                  throw new AssertionError("fix me!");
+                } else if (request.dimension().endsWith(".taxonomy")) {
+                  //if (true) throw new RuntimeException("fix me! " + request.dimension() + "; " + state.facetsConfig.getDimConfig(request.dimension()).indexFieldName);
```
Forgotten commented code?

epotyom replied:
Removed, sorry for the mess!

stefanvodita reviewed (approved):
I left a few more small comments. Thank you for making the changes, great to see this change coming along!

stefanvodita commented:
@mikemccand - did you have more feedback or should we merge?

epotyom commented:
Just a demo of how we can change SearchTask after lucene PR https://github.com/apache/lucene/pull/14237 is merged: https://github.com/epotyom/luceneutil/pull/1/files . Maybe we merge https://github.com/apache/lucene/pull/14237 first, and then apply the change here? WDYT @stefanvodita ?

stefanvodita commented:
Sure, sounds good to me if that's how you'd like to proceed.

mikemccand commented:
> @mikemccand - did you have more feedback or should we merge?

+1 please merge!  Exciting!

epotyom commented:
Before recent changes the test showed regression for some tasks, depending on hardware:
```
                            TaskQPS post_collection_facets      StdDevQPS during_collection_facets      StdDev                Pct diff p-value
         AndHighMedDayTaxoFacets      131.09      (3.8%)       40.94      (0.6%)  -68.8% ( -70% -  -66%) 0.000
     BrowseRandomLabelTaxoFacets        1.94      (1.9%)        0.89      (0.9%)  -54.2% ( -55% -  -52%) 0.000
        AndHighHighDayTaxoFacets       27.37      (3.6%)       14.02      (5.2%)  -48.8% ( -55% -  -41%) 0.000
            MedTermDayTaxoFacets       28.55      (5.4%)       19.98      (7.7%)  -30.0% ( -40% -  -17%) 0.000
          OrHighMedDayTaxoFacets        2.77      (4.2%)        3.33      (2.6%)   19.9% (  12% -   27%) 0.000
       BrowseDayOfYearTaxoFacets        2.76      (3.9%)        4.04      (9.6%)   46.2% (  31% -   62%) 0.000
            BrowseDateTaxoFacets        2.76      (3.8%)        4.04     (11.0%)   46.3% (  30% -   63%) 0.000
           BrowseMonthTaxoFacets        2.30      (4.1%)        3.53     (14.5%)   53.3% (  33% -   74%) 0.000
```

An explanation for BrowseRandomLabelTaxoFacets regression is in [this comment](https://github.com/mikemccand/luceneutil/pull/325#issuecomment-2580729914).

Isolating other regressions was more challenging. The problem is that when I remove all categories except for AndHighMedDayTaxoFacets from the task file, two things happen:
1. Throughput for the category for baseline (post_collection_facets) changes dramatically
2. There is no regression anymore!

The first issue is caused by random selection of one task per category - thanks @mikemccand for explaining how tasks work!

The second issue disappears when I change `groupByCat` from `false` to `true` in [SearchPerfTest.java](https://github.com/mikemccand/luceneutil/blob/21726c39502128ac70023a94f281f2e92ee98525/src/main/perf/SearchPerfTest.java#L628). The reason it helps is that for `during_collection_facets` (candidate) most of the work is done in searcher thread pool, while for `post_collection_facets` (baseline) facets are computed in the tasks thread pool. As a result, running queries from different categories in parallel makes per category results incomparable, because they compete for searcher threads differently. To reduce time when tasks from different categories impact each other, I've added  parameters to `SearchPerfTest`, `benchUtil`, `competition`, `runFacets` to make `groupByCat` configurable, and increased `taskRepeatCount`.

I've also applied commits from [this comment](https://github.com/mikemccand/luceneutil/pull/325#issuecomment-2665470730) to the PR to make the changes available for review, and changed title/description to "pending https://github.com/apache/lucene/pull/14237" 

Test results with all the change:

```
python3 src/python/runFacets.py -source facetsWikimediumAll
```

```
lscpu
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              48

                            TaskQPS post_collection_facets      StdDevQPS during_collection_facets      StdDev                Pct diff p-value
     BrowseRandomLabelTaxoFacets        2.45      (5.8%)        0.75      (0.3%)  -69.3% ( -71% -  -67%) 0.000
            BrowseDateTaxoFacets        3.13      (1.1%)        3.63      (2.9%)   16.0% (  11% -   20%) 0.000
       BrowseDayOfYearTaxoFacets        3.68      (0.6%)        4.32      (1.6%)   17.3% (  15% -   19%) 0.000
           BrowseMonthTaxoFacets        2.64      (1.5%)        3.12      (3.1%)   18.1% (  13% -   22%) 0.000
        AndHighHighDayTaxoFacets        9.70      (2.4%)       12.56      (1.6%)   29.5% (  24% -   34%) 0.000
         AndHighMedDayTaxoFacets       26.33      (1.4%)       36.44      (2.7%)   38.4% (  33% -   43%) 0.000
          OrHighMedDayTaxoFacets        3.69      (3.3%)        5.53      (2.8%)   50.1% (  42% -   58%) 0.000
            MedTermDayTaxoFacets       15.80      (3.4%)       31.02      (4.5%)   96.4% (  85% -  107%) 0.000
```

```
sysctl -a | grep machdep.cpu
machdep.cpu.cores_per_package: 12
machdep.cpu.core_count: 12
machdep.cpu.logical_per_package: 12
machdep.cpu.thread_count: 12
machdep.cpu.brand_string: Apple M3 Pro

                            TaskQPS post_collection_facets      StdDevQPS during_collection_facets      StdDev                Pct diff p-value
     BrowseRandomLabelTaxoFacets        7.25      (4.7%)        1.71      (0.3%)  -76.4% ( -77% -  -74%) 0.000
       BrowseDayOfYearTaxoFacets        9.30      (0.7%)        9.84      (6.5%)    5.9% (  -1% -   13%) 0.000
            BrowseDateTaxoFacets        7.41      (1.1%)        8.46      (6.5%)   14.2% (   6% -   22%) 0.000
           BrowseMonthTaxoFacets        6.20      (6.6%)        7.44     (13.5%)   19.9% (   0% -   42%) 0.000
        AndHighHighDayTaxoFacets       48.05      (4.9%)       64.86      (8.2%)   35.0% (  20% -   50%) 0.000
          OrHighMedDayTaxoFacets        3.14      (1.9%)        4.45      (5.2%)   41.5% (  33% -   49%) 0.000
         AndHighMedDayTaxoFacets       63.83      (4.0%)       93.22     (10.1%)   46.1% (  30% -   62%) 0.000
            MedTermDayTaxoFacets       66.62      (2.3%)      103.72      (5.8%)   55.7% (  46% -   65%) 0.000
```

```
sysctl -a | grep machdep.cpu
machdep.cpu.cores_per_package: 8
machdep.cpu.core_count: 8
machdep.cpu.logical_per_package: 8
machdep.cpu.thread_count: 8
machdep.cpu.brand_string: Apple M3

                            TaskQPS post_collection_facets      StdDevQPS during_collection_facets      StdDev                Pct diff p-value
     BrowseRandomLabelTaxoFacets        7.24      (4.8%)        2.24      (2.0%)  -69.1% ( -72% -  -65%) 0.000
          OrHighMedDayTaxoFacets       19.00      (4.5%)       22.38      (9.5%)   17.8% (   3% -   33%) 0.000
         AndHighMedDayTaxoFacets      121.83      (4.2%)      146.47      (4.8%)   20.2% (  10% -   30%) 0.000
       BrowseDayOfYearTaxoFacets        8.45      (6.6%)       10.16      (7.7%)   20.3% (   5% -   36%) 0.000
            BrowseDateTaxoFacets        6.86      (5.9%)        8.30      (5.2%)   21.0% (   9% -   34%) 0.000
        AndHighHighDayTaxoFacets       13.61      (3.8%)       17.57      (6.6%)   29.1% (  17% -   41%) 0.000
           BrowseMonthTaxoFacets        5.87      (6.3%)        7.78      (9.1%)   32.4% (  16% -   50%) 0.000
            MedTermDayTaxoFacets       79.93      (3.5%)      131.18     (10.6%)   64.1% (  48% -   80%) 0.000
```

stefanvodita commented:
Thank you for doing that investigation @epotyom. Great to see those corrected results!

msokolov commented:
yes, thanks for getting to the bottom of this! And I think you started this whole thread off with an explanation of the one regression (for BrowseRandomLabelTaxoFacets`), so I think we have a convincing story that explains these results now.

epotyom commented:
I think it is ready for merging - there are no conflicts, and the script still works after rebasing (and applying fixes from https://github.com/mikemccand/luceneutil/pull/346 and https://github.com/mikemccand/luceneutil/pull/321 ):

```
python3.7 src/python/runFacets.py -source facetsWikimediumAll

                            TaskQPS post_collection_facets      StdDevQPS during_collection_facets      StdDev                Pct diff p-value
     BrowseRandomLabelTaxoFacets        2.45      (6.6%)        0.74      (0.2%)  -69.6% ( -71% -  -67%) 0.000
            BrowseDateTaxoFacets        3.10      (0.7%)        3.32      (2.2%)    7.2% (   4% -   10%) 0.000
       BrowseDayOfYearTaxoFacets        3.61      (1.2%)        3.98      (3.4%)   10.0% (   5% -   14%) 0.000
           BrowseMonthTaxoFacets        2.55      (3.3%)        2.82      (3.8%)   10.9% (   3% -   18%) 0.000
         AndHighMedDayTaxoFacets       77.67      (5.5%)       99.69      (2.5%)   28.3% (  19% -   38%) 0.000
          OrHighMedDayTaxoFacets        5.89      (2.1%)        7.96      (2.4%)   35.2% (  30% -   40%) 0.000
        AndHighHighDayTaxoFacets       22.99      (3.4%)       34.60      (4.1%)   50.5% (  41% -   60%) 0.000
            MedTermDayTaxoFacets       23.29      (3.1%)       49.10      (5.4%)  110.8% (  99% -  123%) 0.000
```

mikemccand commented:
Yay, thank you everyone!  Especially @epotyom for the PR and the hard deep digging on the tricky regressions.

Also, these are quite astounding performance gains by counting during hit collection instead of 2nd pass after hit collection.

Should we open a follow-on issue to add tasks to `wikinightly.tasks` so we test "during collection faceting" in nightly benchy?
